pkg/medusa/hostmap.go

```diff
@@ -0,0 +1,206 @@
+package medusa
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
+	k8ssandraapi "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	medusaapi "github.com/k8ssandra/k8ssandra-operator/apis/medusa/v1alpha1"
+	cassandrapkg "github.com/k8ssandra/k8ssandra-operator/pkg/cassandra"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type HostName string
+
+type HostDNSOrIP string
+
+// HostMapping is a type that reflects the mapping of source IP address from the Medusa backup to the target host name obtained from looking at the k8s
+// statefulsets that comprise the Cassandra racks.
+type HostMapping struct {
+	Source HostDNSOrIP
+	Target HostName
+}
+type HostMappingSlice []HostMapping
+
+type mappable interface {
+	ToSourceTargetMap() map[HostDNSOrIP]HostName
+	ToTargetSourceMap() map[HostName]HostDNSOrIP
+}
+
+// Transform HostMappingSlide into a map with source IPs as keys and Target IPs as values.
+func (m HostMappingSlice) ToSourceTargetMap() map[HostDNSOrIP]HostName {
+	out := make(map[HostDNSOrIP]HostName)
+	for _, i := range m {
+		out[i.Source] = i.Target
+	}
+	return out
+}
+
+// Transform HostMappingSlide into a map with target IPs as keys and source IPs as values.
+func (m HostMappingSlice) ToTargetSourceMap() map[HostName]HostDNSOrIP {
+	out := make(map[HostName]HostDNSOrIP)
+	for _, i := range m {
+		out[i.Target] = i.Source
+	}
+	return out
+}
+
+// Transform map keyed by target IP with source IP values into HostMappingSlide
+func FromTargetSourceMap(m map[HostName]HostDNSOrIP) HostMappingSlice {
+	out := HostMappingSlice{}
+	for k, v := range m {
+		out = append(out, HostMapping{
+			Source: v,
+			Target: k,
+		})
+	}
+	return out
+}
+
+// Transform map keyed by source IP with target IP values into HostMappingSlide
+func FromSourceTargetMap(m map[HostDNSOrIP]HostName) HostMappingSlice {
+	out := HostMappingSlice{}
+	for k, v := range m {
+		out = append(out, HostMapping{
+			Source: k,
+			Target: v,
+		})
+	}
+	return out
+}
+
+func filterBackupsByName(name string, backups []*BackupSummary) (*BackupSummary, error) {
+	for _, i := range backups {
+		if i.BackupName == name {
+			return i, nil
+		}
+	}
+	return nil, errors.New("could not find named backup")
+}
+
+type NodeLocation struct {
+	Rack string
+	DC   string
+}
+
+type backupGetter interface {
+	GetBackups(ctx context.Context) ([]*BackupSummary, error)
+}
+
+// getBackupRackIPs gets a map of racks to IPs from a Medusa CassandraBackup k8s object.
```

Comment:
> You may want to trade "IPs" for "Host" in the following functions since we won't be dealing with just IPs. We can have hostnames as well.

Agreed and done. This fell out of sync when I was making changes to accomodate IPs in the source.

> Also the comment repeats (incorrectly) the name of the function, which is probably not something that's desirable: if you change the name of the function (as you did) then you need to change it in the comment as well.

It is indeed unfortunate, however it is [required](https://go.dev/blog/godoc) for godoc to pick up the comment as documenting the function:

> to document a type, variable, constant, function, or even a package, write a regular comment directly preceding its declaration, with no intervening blank line. 

Whether we strictly need to document a private function I'm not sure, but I'm trying to make a habit of documenting as much as possible.